core/crypto/kem/adapter/kem.go

```diff
@@ -0,0 +1,249 @@
+		return nil, kem.ErrCiphertextSize
+	}
+	pk, err := a.UnmarshalBinaryPublicKey(ct)
+	if err != nil {
+		return nil, err
+	}
+	ss := a.nike.DeriveSecret(sk.(*PrivateKey).privateKey, pk.(*PublicKey).publicKey)
+	ss2 := h(a.nike.PublicKeySize(), ss, sk.Public().(*PublicKey).publicKey.Bytes(), pk.(*PublicKey).publicKey.Bytes())
+	return ss2, nil
+}
+
+// Unmarshals a PublicKey from the provided buffer.
+func (a *Scheme) UnmarshalBinaryPublicKey(b []byte) (kem.PublicKey, error) {
+	pubkey, err := a.nike.UnmarshalBinaryPublicKey(b)
+	if err != nil {
+		return nil, err
+	}
+	return &PublicKey{
+		publicKey: pubkey,
+		scheme:    a,
+	}, nil
+}
+
+// Unmarshals a PrivateKey from the provided buffer.
+func (a *Scheme) UnmarshalBinaryPrivateKey(b []byte) (kem.PrivateKey, error) {
+	privkey, err := a.nike.UnmarshalBinaryPrivateKey(b)
+	if err != nil {
+		return nil, err
+	}
+	return &PrivateKey{
+		privateKey: privkey,
+		scheme:     a,
+	}, nil
+}
+
+// Size of encapsulated keys.
+func (a *Scheme) CiphertextSize() int {
+	return a.nike.PublicKeySize()
+}
+
+// Size of established shared keys.
+func (a *Scheme) SharedKeySize() int {
+	return a.nike.PublicKeySize()
+}
+
+// Size of packed private keys.
+func (a *Scheme) PrivateKeySize() int {
+	return a.nike.PrivateKeySize()
+}
+
+// Size of packed public keys.
+func (a *Scheme) PublicKeySize() int {
+	return a.nike.PublicKeySize()
+}
+
+// DeriveKeyPair deterministicallly derives a pair of keys from a seed.
+// Panics if the length of seed is not equal to the value returned by
+// SeedSize.
+func (a *Scheme) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
+	if len(seed) != a.SeedSize() {
+		panic(kem.ErrSeedSize)
+	}
+	h := sha3.NewShake256()
+	_, _ = h.Write(seed)
+	pk, sk, err := a.nike.GenerateKeyPairFromEntropy(h)
+	if err != nil {
+		panic(err)
+	}
+	return &PublicKey{
+			publicKey: pk,
+			scheme:    a,
+		}, &PrivateKey{
+			privateKey: sk,
+			scheme:     a,
+		}
+}
+
+// Size of seed used in DeriveKey
+func (a *Scheme) SeedSize() int {
+	return a.nike.PublicKeySize()
+}
+
+// EncapsulateDeterministically generates a shared key ss for the public
+// key deterministically from the given seed and encapsulates it into
+// a ciphertext ct. If unsure, you're better off using Encapsulate().
+func (a *Scheme) EncapsulateDeterministically(pk kem.PublicKey, seed []byte) (
+	[]byte, []byte, error) {
+	if len(seed) != a.EncapsulationSeedSize() {
+		return nil, nil, kem.ErrSeedSize
+	}
+	pub, ok := pk.(*PublicKey)
+	if !ok || pub.scheme != a {
+		return nil, nil, kem.ErrTypeMismatch
+	}
+
+	pk2, sk2 := a.DeriveKeyPair(seed)
+	ss := a.nike.DeriveSecret(sk2.(*PrivateKey).privateKey, pub.publicKey)
+	ss2 := h(a.nike.PublicKeySize(), ss, pub.publicKey.Bytes(), pk2.(*PublicKey).publicKey.Bytes())
+	ct, _ := pk2.MarshalBinary()
+	return ct, ss2, nil
+}
+
+// Size of seed used in EncapsulateDeterministically().
+func (a *Scheme) EncapsulationSeedSize() int {
+	return a.nike.PublicKeySize()
```

Comment:
Shouldn't this be `SeedSize` ?

It looks like it calls this function:
```go
// DeriveKeyPair deterministicallly derives a pair of keys from a seed.
// Panics if the length of seed is not equal to the value returned by
// SeedSize.
func (a *Scheme) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
	if len(seed) != a.SeedSize() {
```

I guess for testing whether we got this right we could have a `HeteregenousLengthTestScheme` with distinct lengths for each so we can identify places where we screw this up before slipups make it into the rest of our codebase?

Reply:
fixed. set it to 32